components/support/migration/src/main/java/mozilla/components/support/migration/AddonMigration.kt

```diff
@@ -0,0 +1,135 @@
+     */
+    sealed class Success : AddonMigrationResult() {
+        /**
+         * No add-ons installed to migrate.
+         */
+        object NoAddons : Success()
+
+        /**
+         * Successfully migrated add-ons.
+         */
+        data class AddonsMigrated(
+            val migratedAddons: List<WebExtension>
+        ) : Success() {
+            override fun toString(): String {
+                return "Successfully migrated ${migratedAddons.size} add-ons."
+            }
+        }
+    }
+
+    /**
+     * Failure variants of an add-on migration.
+     */
+    sealed class Failure : AddonMigrationResult() {
+        /**
+         * Failed to query installed add-ons via the [Engine].
+         */
+        internal data class FailedToQueryInstalledAddons(val throwable: Throwable) : Failure() {
+            override fun toString(): String {
+                return "Failed to query installed add-ons: ${throwable::class}"
+            }
+        }
+
+        /**
+         * Failed to migrate some or all of the installed add-ons.
+         */
+        internal data class FailedToMigrateAddons(
+            val migratedAddons: List<WebExtension>,
+            val failedAddons: Map<WebExtension, Throwable>
+        ) : Failure() {
+            override fun toString(): String {
+                return "Failed to migrate ${failedAddons.size} of ${failedAddons.size + migratedAddons.size} add-ons."
+            }
+        }
+    }
+}
+
+/**
+ * Helper for migrating add-ons.
+ */
+internal object AddonMigration {
+
+    /**
+     * Performs a migration of all installed add-ons. The only migration step involved
+     * is to make sure we disable all currently unsupported add-ons.
+     */
+    @SuppressWarnings("TooGenericExceptionCaught")
+    suspend fun migrate(engine: Engine): Result<AddonMigrationResult> {
+        try {
+            val installedAddons = getInstalledAddons(engine)
+            if (installedAddons.isEmpty()) {
+                return Result.Success(AddonMigrationResult.Success.NoAddons)
+            }
+
+            val migratedAddons = mutableListOf<WebExtension>()
+            val failures = mutableMapOf<WebExtension, Exception>()
+            installedAddons.forEach { addon ->
+                try {
+                    migratedAddons += migrateAddon(engine, addon)
+                } catch (e: Exception) {
+                    failures[addon] = e
+                }
+            }
+
+            return if (failures.isEmpty()) {
+                Result.Success(AddonMigrationResult.Success.AddonsMigrated(migratedAddons))
+            } else {
+                val failure = AddonMigrationResult.Failure.FailedToMigrateAddons(migratedAddons, failures)
+                Result.Failure(AddonMigrationException(failure))
+            }
+        } catch (e: Exception) {
+            return Result.Failure(AddonMigrationException(AddonMigrationResult.Failure.FailedToQueryInstalledAddons(e)))
```

Comment:
A bit of a nit, but `FailedToQueryInstalledAddons` won't be strictly the right failure if something other than `getInstalledAddons` fails.

Reply:
Yeah, that's fair. There's no need for a catch-all here as I am doing that in `FennecMigrator#migrateAddons` anyway. Refactored.